pymodbus/client/mixin.py

```diff
@@ -66,9 +66,11 @@ def read_coils(
         :param kwargs: (optional) Experimental parameters.
         :raises ModbusException:
         """
-        return self.execute(
+        response = self.execute(
             pdu_bit_read.ReadCoilsRequest(address, count, slave, **kwargs)
         )
+        assert isinstance(response, pdu_bit_read.ReadCoilsResponse)
```

Comment:
This will not work in practical life !! we only get a ReadCoilsResponse if the device responded positively...otherwise we get typically ModBusIOException (or something similar) this is NOT an exception but a message from the device.

Apart from that it is hard to see how that helps on mypy. I can accept it if you remove the assert, but it really needs a comments as to why this is nessecary.

Reply:
> This will not work in practical life !! we only get a ReadCoilsResponse if the device responded positively...otherwise we get typically ModBusIOException (or something similar) this is NOT an exception but a message from the device.

Good point, and it means all the existing type signatures are too narrow.
They should become, for example, `Union[pdu_bit_read.ReadCoilsResponse, exceptions.ModbusException]`



> Apart from that it is hard to see how that helps on mypy. I can accept it if you remove the assert, but it really needs a comments as to why this is nessecary.

The problem is the return type for `execute()` is too broad, and the other methods want _more specific_ types.

```python

def execute(self, request: ModbusRequest) -> ModbusResponse:  # cannot be more specific with static analysis, only at runtime
    raise ModbusException("Pymodbus internal ERROR")

def read_coils(
    self, address: int, count: int = 1, slave: int = 0, **kwargs: Any
) -> pdu_bit_read.ReadCoilsResponse:  # more specific than ModbusResponse
```

I think the best solution is to remove the response type hint for `execute()`.  I did this in https://github.com/pymodbus-dev/pymodbus/pull/1356/commits/636e9a779774302bfcbf0d82a4542c11965f56a4 but you disagreed  :).  I am trying alternative solutions, but don't like any of them (hence 'Draft' status)

Reply:
exceptions.ModbusException is a real exception and not a modbus Message. Have a look at client_calls.py in examples to see what can be returned (or raised).

Something is not right, in home assistant we often use broad classes (like e.g. sensor) that are then narrowed down when being used (like e.g. a binary_sensor). However if the problems goes away by removing the return type from execute() then by all means. 

The alternative is to add a cast to the execute() in each method.